### ISSUE TITLE
matmul_kernel_persistent runs fp32 inputs through TF32 — silent precision loss vs torch.mm

### ISSUE THREAD

**Issue by chenwenxiaolive** [thinking-machines-lab/batch_invariant_ops#23]
### Summary

`set_batch_invariant_mode(True)` overrides `aten::mm` / `aten::addmm` with `matmul_kernel_persistent`. That kernel calls

```python
accumulator = tl.dot(a, b, accumulator)
```

without `input_precision`. On NVIDIA Tensor-Core GPUs, Triton's `tl.dot` **defaults to TF32 for fp32 inputs** ([docs](https://triton-lang.org/main/python-api/generated/triton.language.dot.html)). TF32 has a 10-bit mantissa, so integers are exact only up to 2¹¹ = 2048.

PyTorch's own `torch.mm` runs fp32 in IEEE by default (`torch.backends.cuda.matmul.allow_tf32 == False` since PyTorch 1.12). So enabling batch-invariant mode **silently downgrades fp32 matmul precision** from IEEE to TF32 — an accuracy change unrelated to (and not required by) batch invariance.

### Reproduction (pure tensor, no model)

```python
import torch
from batch_invariant_ops import set_batch_invariant_mode

device = "cuda"
pos  = torch.arange(8192, device=device, dtype=torch.float32)  # every value exact in fp32
ones = torch.ones(1, 1, device=device, dtype=torch.float32)

ref = torch.mm(ones, pos[None, :])                 # mode OFF: IEEE fp32
assert torch.equal(ref[0], pos)                    # exact

with set_batch_invariant_mode(True):               # mode ON: matmul_kernel_persistent
    out = torch.mm(ones, pos[None, :])

print("mismatches:", int((out[0] != pos).sum()))   # 4096
print("out[2049] =", out[0, 2049].item())          # 2048.0  (expected 2049.0)
print("out[4099] =", out[0, 4099].item())          # 4096.0  (expected 4099.0)
print("out[8191] =", out[0, 8191].item())          # 8188.0  (expected 8191.0)
```

Output:
```
mismatches: 4096
out[2049] = 2048.0
out[4099] = 4096.0
out[8191] = 8188.0
```

The error pattern matches TF32 ULP exactly: exact in [0, 2048), ULP 2 in [2048, 4096) (odd integers lost → 1024 wrong), ULP 4 in [4096, 8192) (3/4 lost → 3072 wrong), total 4096.

### Why this matters: rotary embeddings (RoPE)

The standard rotary embedding phase is `positions ⊗ inv_freq`, an fp32 matmul. Computed through this kernel, both the positions and the (sub-1.0) `inv_freq` values are TF32-truncated. With **real Llama-3.1 params** (`head_dim=128`, `rope_theta=500000`), comparing `positions[:, None] @ inv_freq[None, :]` through `matmul_persistent` vs `torch.mm` over a 4096-token context:

```
RoPE cos() max-abs-error vs IEEE reference:
  positions    0–2048 : 4.76e-01     # inv_freq truncation (all positions)
  positions 2048–4096 : 1.35e+00     # + integer-position truncation beyond 2048
```

i.e. the rotary phase is corrupted by up to ~1.35 in `cos`. (With `input_precision="ieee"` the error is exactly `0.0`, bitwise.)

Note: this repo overrides only `aten::{mm, addmm}`, so the 3-D `inv_freq @ position_ids` form used by HuggingFace models is **not** intercepted here. But downstream batch-invariant inference integrations that extend the same op-override approach to `aten::{matmul, bmm}` (e.g. vLLM) route a model's actual rotary through equivalent kernels and get long-context RoPE corruption beyond position 2048 (filing separately against vLLM).

### Fix

One line in `matmul_kernel_persistent`:

```diff
-            accumulator = tl.dot(a, b, accumulator)
+            accumulator = tl.dot(a, b, accumulator, input_precision="ieee")
```

`input_precision="ieee"` is a **no-op for bf16/fp16** (TF32 only applies to fp32), so the typical bf16 GEMM path is unchanged. PR with a regression test to follow.
